certbot-dns-cloudflare/certbot_dns_cloudflare/dns_cloudflare.py

```diff
@@ -0,0 +1,202 @@
+"""DNS Authenticator for Cloudflare."""
+import logging
+
+import zope.interface
+
+from certbot import errors
+from certbot import interfaces
+
+from certbot.plugins import dns_common
+
+import CloudFlare
+
+logger = logging.getLogger(__name__)
+
+ACCOUNT_URL = 'https://www.cloudflare.com/a/account/my-account'
+
+
+@zope.interface.implementer(interfaces.IAuthenticator)
+@zope.interface.provider(interfaces.IPluginFactory)
+class Authenticator(dns_common.DNSAuthenticator):
+    """DNS  Authenticator for Cloudflare
+
+    This Authenticator uses the Cloudflare API to fulfill a dns-01 challenge.
+    """
+
+    description = 'Obtain certs using a DNS TXT record (if you are using Cloudflare for DNS).'
+
+    ttl = 120
+    credentials = None
+
+    def __init__(self, *args, **kwargs):
+        super(Authenticator, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def add_parser_arguments(cls, add):
+        super(Authenticator, cls).add_parser_arguments(add)
+        add('credentials', help='Cloudflare credentials file.')
+        # TODO: Document format (without calling `self.dest`?!)
+
+    def more_info(self):  # pylint: disable=missing-docstring,no-self-use
+        return 'This plugin configures a DNS TXT record to respond to a dns-01 challenge using ' + \
+               'the Cloudflare API.'
+
+    def _setup_credentials(self):
+        self.credentials = self._configure_credentials(
+            'credentials',
+            'Cloudflare credentials file',
+            {
+                'email': 'email address associated with Cloudflare account',
+                'api-key': 'API key for Cloudflare account, obtained from {0}'.format(ACCOUNT_URL)
+            }
+        )
+
+    def _perform(self, domain, validation_name, validation):
+        self._get_cloudflare_client().add_txt_record(domain, validation_name, validation, self.ttl)
+
+    def _cleanup(self, domain, validation_name, validation):
+        self._get_cloudflare_client().del_txt_record(domain, validation_name, validation)
+
+    def _get_cloudflare_client(self):
+        return _CloudflareClient(self.credentials.conf('email'), self.credentials.conf('api-key'))
```

Comment:
There's little overhead to creating a new client and using a new one for each operation means we don't need to worry about the client (or the underlying library) storing state. If we _always_ create a new client, we know there's never state, making the code easier to reason about.